modules/renter/contractor/contracts.go

```diff
@@ -283,6 +359,132 @@ func (c *Contractor) managedRenew(sc *proto.SafeContract, contractFunding types.
 	return newContract, nil
 }
 
+// managedRenewContract will use the renew instructions to renew a contract,
+// returning the amount of money that was put into the contract for renewal.
+func (c *Contractor) managedRenewContract(renewInstructions fileContractRenewal, currentPeriod types.BlockHeight, allowance modules.Allowance, blockHeight types.BlockHeight) (fundsSpent types.Currency, err error) {
+	// Pull the variables out of the renewal.
+	id := renewInstructions.id
+	amount := renewInstructions.amount
+
+	// Mark the contract as being renewed, and defer logic to unmark it
+	// once renewing is complete.
+	c.mu.Lock()
+	c.renewing[id] = true
+	c.mu.Unlock()
+	defer func() {
+		c.mu.Lock()
+		delete(c.renewing, id)
+		c.mu.Unlock()
+	}()
+
+	// Wait for any active editors and downloaders to finish for this
+	// contract, and then grab the latest revision.
+	c.mu.RLock()
+	e, eok := c.editors[id]
+	d, dok := c.downloaders[id]
+	c.mu.RUnlock()
+	if eok {
+		e.invalidate()
+	}
+	if dok {
+		d.invalidate()
+	}
+
+	// Fetch the contract that we are renewing.
+	oldContract, exists := c.staticContracts.Acquire(id)
+	if !exists {
+		return types.ZeroCurrency, errors.New("contract no longer exists")
+	}
+	// Return the contract if it's not useful for renewing.
+	oldUtility, ok := c.managedContractUtility(id)
+	if !ok || !oldUtility.GoodForRenew {
+		c.log.Printf("Contract %v slated for renew is marked not good for renew %v/%v",
+			id, ok, oldUtility.GoodForRenew)
+		c.staticContracts.Return(oldContract)
+		return types.ZeroCurrency, errors.New("contract is marked not good for renew")
+	}
+
+	// Calculate endHeight for renewed contracts
+	endHeight := currentPeriod + allowance.Period
```

Comment:
there is a function [`contractEndHeight()` ](https://github.com/NebulousLabs/Sia/blob/8da8220708ed647b7fa9e466a0b2ee5355666e47/modules/renter/contractor/contracts.go#L27) which can be called on the `Contractor`.  We should probably use that instead of manually calculating the endHeight in multiple places right?

Reply:
contractEndHeight needs a lock